### ISSUE TITLE
Auth - MFA

### ISSUE THREAD

**Issue by jpmtrabbold** [serverless-stack/sst#2170]
Hello guys, can we please look into incorporating MFA into Auth?

As per this [thread](https://discord.com/channels/983865673656705025/1032451504986140672) (more info in the thread), I'm creating this issue so this can be tracked and prioritised properly in the roadmap.

Most enterprise solutions require MFA, even with the Magic Link flow. Even though MFA can be implemented outside of Auth's implementation, it would still be an incomplete solution, because the JWT is already issued and valid, whereas in an ideal world, the JWT should only be issued when MFA is checked (I think?).

And as an additional thought, maybe we can discuss how [passkeys](https://fidoalliance.org/passkeys/) will affect Auth going forward?

**Comment by jayair** [comment]
Just to clarify, for MFA do you mean verifying a code that's sent through email or SMS. Or do you mean through an app like Google Authenticator?

**Comment by jpmtrabbold** [comment]
Thanks for getting back to me, @jayair! I'm sure there are multiple scenarios, but thinking of the Magic Link flow:

- the Magic Link would be the primary authentication factor - the user receives the link either by e-mail or SMS;
- to be MFA "compliant", we would need at least a second authentication factor that doesn't rely on the same method used by the primary factor. E.g. if the magic link were sent by e-mail, then the second factor could be a code, either from an Authenticator app or by SMS;

I'm happy to be challenged on that. From my research, I couldn't tell if the Magic Link method by itself could be considered as MFA or not.

**Comment by jayair** [comment]
Yeah that makes sense. Let me see what @thdxr has to say on this.

**Comment by thdxr** [comment]
Hey gave this some thought, I think realistically MFA is a bit out of scope in the near term. We need to cover other providers before we think about something like MFA which likely will be very stateful.  Would probably suggest implementing it yourself for the time being

**Comment by joaothegapportal** [comment]
@thdxr I understand you guys' priorities. Maybe 2 things to be considered:
- if you want to appeal to enterprise-grade applications, mfa is a must. Or at least support passkeys. Something that gives that additional layer of security;
- mfa would be as stateful as any other piece of Auth - but then you defer the stateful bits to the app's implementation - both setting and getting - via hooks, as you are already doing;

**Comment by jpmtrabbold** [comment]
@thdxr @jayair Just an update on this. I've noticed that people have been talking about [WebAuthn](https://webauthn.io/) a lot more now that Apple implemented Passkeys on iOS 16/Mac Ventura. It seems like the safe and user-friendly way of getting rid of passwords because it is considered, in itself, a two-factor authentication method (device + biometrics or device + platform password). Is that something of interest for SST Auth?

**Comment by thdxr** [comment]
We are transition this repo to https://github.com/sst/v2 and are closing this issue for now. If it is still relevant please reopen it there - sorry for the trouble
